File: CardGameTracker.Core/Models/Wizard/Score.cs
namespace CardGameTracker.Models.Wizard;

public class Score
{
    public Score()
    {
        Bid = 0;
        TricksTaken = null;
    }

    public Score(int bid, int tricksTaken)
    {
        Bid = bid;
        TricksTaken = tricksTaken;
    }

    private int _bid;
    public int Bid
    {
        get => _bid;
        set
        {
            if (value < 0)
            {
                _bid = 0;
            }
            else
            {
                _bid = value;
            }
            CalculateScore();
        }
    }
    private int? _tricksTaken;
    public int? TricksTaken
    {
        get => _tricksTaken;
        set
        {
            if (value < 0)
            {
                _tricksTaken = 0;
            }
            else
            {
                _tricksTaken = value;
            }
            CalculateScore();
        }
    }

    public int? CalculateScore()
    {

        if (TricksTaken == null)
        {
            CalculatedScore = null;
        }
        else if (Bid == TricksTaken)
        {
            CalculatedScore = 20 + Bid * 10;
        }
        else
        {
            CalculatedScore = Math.Abs(Bid - TricksTaken.Value) * -10;
        }

        return CalculatedScore;
    }

    public int? CalculatedScore { get; private set; }
}

File: card-game-tracker-react/src/models/wizard/score.ts
import { computed, effect, signal } from "@preact/signals-react";

interface IScoreData {
    bid: number;
    tricksTaken?: number;
}

class Score {

    constructor(score?: IScoreData) {
        if (score !== undefined) {
            this.Bid = score.bid;
            this.TricksTaken = score.tricksTaken;
        }
    }

    private count = 0;

    bid = signal(0);
    peekBid() { return this.bid.peek(); }
    get Bid() { return this.bid.value; }
    set Bid(value) { this.bid.value = value; }

    private tricksTaken = signal<number | undefined>(undefined);
    peekTricks() { return this.tricksTaken.peek(); }
    get TricksTaken() { return this.tricksTaken.value; }
    set TricksTaken(value) { this.tricksTaken.value = value; }

    score = computed(() =>
        (this.Bid === this.TricksTaken)
            ? 20 + (this.Bid * 10)
            : (this.TricksTaken === undefined)
                ? 0
                : Math.abs(this.Bid - this.TricksTaken) * -10);

    get ComputedScore() { return this.score.value; }

}

export default Score;
export type { IScoreData };